feat: add `blas/ext/index-of`

headlessNode opened the pull request:
Progresses #2656.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `blas/ext/index-of`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2656 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add blas/ext/index-of

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/base.js:
```diff
@@ -0,0 +1,109 @@
+* @name indexOf
+* @type {Function}
+* @param {ndarrayLike} x - input ndarray
+* @param {(ndarrayLike|number)} searchElement - search element
```
```suggestion
* @param {(ndarrayLike|*)} searchElement - search element
```

This is a recurring pattern. The search element should not be limited to numbers.

kgryte replied:
Also, this is wrong. As in `cusum`, at this stage, all arguments should be ndarrays. Not scalars. See docs for https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/ndarray/base/unary-reduce-strided1d-dispatch-factory.

kgryte replied:
Also, the type should be `ndarray`. Again, by this point, we've done the argument juggling and what we should be passing down are `ndarray` instances.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/base.js:
```diff
@@ -0,0 +1,109 @@
+* @type {Function}
+* @param {ndarrayLike} x - input ndarray
+* @param {(ndarrayLike|number)} searchElement - search element
+* @param {(ndarrayLike|number)} [fromIndex] - index from which to begin searching
```
```suggestion
* @param {ndarray} fromIndex - indices from which to begin searching
```

kgryte replied:
Same comment.

kgryte replied:
Also, by this point, `fromIndex` should always be provided. In your lower-level implementations, where would the default value of `fromIndex` be supplied? TL;DR: there isn't such a lower-level. In `blas/ext/base/ndarray/gindex-of`, we always assume a third ndarray: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/blas/ext/base/ndarray/gindex-of. This was a package you just worked on.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/base.js:
```diff
@@ -0,0 +1,109 @@
+* @param {IntegerArray} [options.dims] - list of dimensions over which to perform operation
+* @param {string} [options.dtype] - output ndarray data type
+* @throws {TypeError} first argument must be an ndarray-like object
+* @throws {TypeError} search element argument must be either an ndarray-like object or a scalar value
```
```suggestion
* @throws {TypeError} second argument must be either an ndarray-like object
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/base.js:
```diff
@@ -0,0 +1,109 @@
+* @param {string} [options.dtype] - output ndarray data type
+* @throws {TypeError} first argument must be an ndarray-like object
+* @throws {TypeError} search element argument must be either an ndarray-like object or a scalar value
+* @throws {TypeError} from index argument must be either an ndarray-like object or a scalar value
```
```suggestion
* @throws {TypeError} third argument must be either an ndarray-like object
```

kgryte replied:
We try to avoid specifically naming arguments like you've done in order to avoid 1) copy-paste mistakes and 2) the potential for future drift when code is refactored, but docs are not. The exception we make is `options` which can "float" around to different argument positions depending on the API.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/base.js:
```diff
@@ -0,0 +1,109 @@
+* var x = new ndarray( 'float64', xbuf, sh, sx, ox, 'row-major' );
+*
+* // Find index:
+* var out = indexOf( x, 4.0 );
```
This is incorrect. You should be passing down a zero-dimensional ndarray if you want a "scalar" argument.

headlessNode replied:
Yeah, Initially I used `scalar2ndarray` but had an `data property undefined` error.

kgryte replied:
Also incorrect in that there should be a `fromIndex` ndarray.

kgryte replied:
> Yeah, Initially I used scalar2ndarray but had an data property undefined error.

Yes, likely because you were not passing down a `fromIndex` ndarray.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+* @name indexOf
+* @type {Function}
+* @param {ndarrayLike} x - input ndarray
+* @param {(ndarrayLike|number)} searchElement - search element
```
```suggestion
* @param {(ndarrayLike|*)} searchElement - search element
```

kgryte replied:
Please fix this here and everywhere else. Having to correct this is getting repetitive.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+* @type {Function}
+* @param {ndarrayLike} x - input ndarray
+* @param {(ndarrayLike|number)} searchElement - search element
+* @param {(ndarrayLike|number)} [fromIndex] - index from which to begin searching
```
```suggestion
* @param {(ndarrayLike|integer)} [fromIndex] - index from which to begin searching
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+* @param {boolean} [options.keepdims=false] - boolean indicating whether the reduced dimensions should be included in the returned ndarray as singleton dimensions
+* @param {string} [options.dtype] - output ndarray data type
+* @throws {TypeError} first argument must be an ndarray-like object
+* @throws {TypeError} search element argument must be either an ndarray-like object or a numeric value
```
Incorrect.

kgryte replied:
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+* @param {string} [options.dtype] - output ndarray data type
+* @throws {TypeError} first argument must be an ndarray-like object
+* @throws {TypeError} search element argument must be either an ndarray-like object or a numeric value
+* @throws {TypeError} from index argument must be either an ndarray-like object or a numeric value
```
```suggestion
* @throws {TypeError} third argument must be either an ndarray-like object or an integer
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+			return base( x, searchElement, broadcastScalar( 0, DEFAULT_DTYPE, [], ord ) ); // eslint-disable-line max-len
+		}
+		// Case: indexOf( x, searchElement_scalar )
+		if ( isNumber( searchElement ) ) {
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+		if ( isNumber( searchElement ) ) {
+			return base( x, broadcastScalar( searchElement, dt, [], ord ), broadcastScalar( 0, DEFAULT_DTYPE, [], ord ) ); // eslint-disable-line max-len
+		}
+		throw new TypeError( format( 'invalid argument. Second argument must be either an ndarray or a scalar value. Value: `%s`.', searchElement ) );
```
This error should not be possible. When `searchElement` is not an ndarray-like object, it should always be considered a "scalar".

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+	// Resolve input ndarray meta data:
+	dt = getDType( x );
+	ord = getOrder( x );
+
```
Explicitly handle the case where `nargs < 2` first. Right now, this gets handled with `nargs < 4` below which just leads to incorrect error messages.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+		throw new TypeError( format( 'invalid argument. Second argument must be either an ndarray or a scalar value. Value: `%s`.', searchElement ) );
+	}
+	// Case: indexOf( x, searchElement, ??? )
+	else if ( nargs < 4 ) { // nargs < 4
```
```suggestion
	else if ( nargs === 3 ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+			// As the operation is performed across all dimensions, `se` & `fromIndex` is assumed to be a zero-dimensional ndarray...
+			return base( x, searchElement, fromIndex );
+		}
+		if ( isNumber( searchElement ) && isNumber( fromIndex ) ) {
```
`fromIndex` should not be allowed to be a fractional value. In what use case does `fromIndex = 1.5` make sense?

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+			return base( x, broadcastScalar( searchElement, dt, [], ord ), broadcastScalar( fromIndex, DEFAULT_DTYPE, [], ord ) ); // eslint-disable-line max-len
+		}
+		if ( isndarrayLike( searchElement ) && isNumber( fromIndex ) ) {
+			// As the operation is performed across all dimensions, `se` & `fromIndex` is assumed to be a zero-dimensional ndarray...
```
What is `se`?

kgryte replied:
Also, this comment is incorrect based on what happens in the subsequent line.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+			// As the operation is performed across all dimensions, `se` & `fromIndex` is assumed to be a zero-dimensional ndarray...
+			return base( x, searchElement, broadcastScalar( fromIndex, DEFAULT_DTYPE, [], ord ) ); // eslint-disable-line max-len
+		}
+		if ( isNumber( searchElement ) && isndarrayLike( fromIndex ) ) {
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+	// Case: indexOf( x, searchElement, ??? )
+	else if ( nargs < 4 ) { // nargs < 4
+		// Case: indexOf( x, searchElement, fromIndex )
+		if ( isndarrayLike( searchElement ) && isndarrayLike( fromIndex ) ) {
```
You are doing repeated `is*` checks. You need to consolidate them. Figure out the input types first to avoid potentially expensive repeated branch checks.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+			searchElement = broadcastScalar( searchElement, dt, sh, getOrder( x ) ); // eslint-disable-line max-len
+			fromIndex = broadcastScalar( 0, DEFAULT_DTYPE, sh, ord ); // eslint-disable-line max-len
+		} else {
+			throw new TypeError( format( 'invalid argument. Second argument must be either an ndarray or a numeric scalar value. Value: `%s`.', searchElement ) );
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,204 @@
+	opts = options;
+
+	// Case: indexOf( x, searchElement_ndarray, fromIndex_ndarray, opts )
+	if ( isndarrayLike( searchElement ) && isndarrayLike( fromIndex ) ) {
```
Same comment. You are doing repeated brand checks. Handle each argument separately. Normalize each to a broadcasted ndarray. Do not do what you are doing here with handling combinations. This just leads to noise and repeated logic.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/assign.js:
```diff
@@ -0,0 +1,228 @@
+	dt = getDType( x );
+	ord = getOrder( x );
+
+	if ( nargs === 3 ) {
```
Essentially all of the same comments from `main.js`.

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: cleanup & refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: refactor & cleanup

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
docs: fix err description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add main tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add missing case

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add assign tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: incorrect strides in test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/README.md:
```diff
@@ -0,0 +1,262 @@
+var x = array( [ 1.0, 2.0, 3.0, 4.0 ] );
+
+var idx = indexOf( x, 2.0, {
+    'dtype': 'generic'
```
```suggestion
    'dtype': 'int32'
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/README.md:
```diff
@@ -0,0 +1,262 @@
+// returns <ndarray>
+
+var dt = dtype( idx );
+// returns 'generic'
```
```suggestion
// returns 'int32'
```

headlessNode committed:
docs: apply suggestions from cide review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/docs/types/index.d.ts:
```diff
@@ -0,0 +1,222 @@
+	* var idx = y.get();
+	* // returns 1
+	*/
+	<T = unknown>( x: InputArray<T>, searchElement: SearchElement<T>, options?: Options ): OutputArray; // NOTE: we lose type specificity here, but retaining specificity would likely be difficult and/or tedious to completely enumerate, as the output ndarray data type is dependent on how `x` interacts with output data type policy and whether that policy has been overridden by `options.dtype`.
```
```suggestion
	<T = unknown>( x: InputArray<T>, searchElement: SearchElement<T>, options?: Options ): OutputArray;
```

headlessNode committed:
chore: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/base.js:
```diff
@@ -0,0 +1,120 @@
+// VARIABLES //
+
+var idtypes0 = dtypes( 'numeric_and_generic' ); // input ndarray
+var idtypes1 = dtypes( 'numeric_and_generic' ); // search element ndarray
```
These sets of dtypes are still incorrect. Why should `bool` be excluded?

headlessNode replied:
~~`bool` will fall through to `generic`, no?~~

headlessNode replied:
Yeah....I think it should be `all` yes?

kgryte replied:
Correct.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/main.js:
```diff
@@ -0,0 +1,164 @@
+	if ( nargs === 2 ) {
+		fromIndex = 0;
+	}
+	// Case: indexOf( x, searchElement, options )
```
What about the case: `indexOf( x, searchElement, fromIndex )`?

headlessNode replied:
We are handling `options` here. For `indexOf( x, searchElement, fromIndex )` there is no `options` and we already intialized `opts = {}`.

So the below checks (before return) on `searchElement` and `fromIndex` handles this case correctly.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/base.js:
```diff
@@ -0,0 +1,120 @@
+
+// VARIABLES //
+
+var idtypes0 = dtypes( 'numeric_and_generic' ); // input ndarray
+var idtypes1 = dtypes( 'numeric_and_generic' ); // search element ndarray
```
```suggestion
var idtypes0 = dtypes( 'all' ); // input ndarray
var idtypes1 = dtypes( 'all' ); // search element ndarray
```

headlessNode committed:
fix: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
test: add missing test case

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
Merge branch 'top-blas-indexof' of https://github.com/headlessNode/stdlib into top-blas-indexof

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/test/test.assign.js:
```diff
@@ -0,0 +1,1662 @@
+
+	function badValue( value ) {
+		return function badValue() {
+			indexOf( x, scalar2ndarray( 2.0 ), y, value, {} );
```
```suggestion
			indexOf( x, scalar2ndarray( 2.0 ), value, y, {} );
```

headlessNode committed:
test: fix arguments

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/README.md:
```diff
@@ -0,0 +1,262 @@
+
+-   **x**: input [ndarray][@stdlib/ndarray/ctor].
+-   **searchElement**: element in an input [ndarray][@stdlib/ndarray/ctor] for which to find an index. May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having a [data type][@stdlib/ndarray/dtypes] same as the data type of the input [ndarray][@stdlib/ndarray/ctor]. If provided a scalar value, the value is cast to the data type of the input [ndarray][@stdlib/ndarray/ctor]. If prov ...
+-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an `integer` or `generic` [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the complement of the shape defined by `o ...
```
```suggestion
-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an `integer` or `generic` [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the complement of the shape defined by `options.dims`. For example, given the input shape `[2, 3, 4]` and `options.dims=[0]`, an [ndarray][@stdlib/ndarray/ctor] containing the index from which to begin searching must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the shape `[3, 4]`. Similarly, when performing the operation over all elements in a provided input [ndarray][@stdlib/ndarray/ctor], an [ndarray][@stdlib/ndarray/ctor] containing the index from which to begin searching must be a zero-dimensional [ndarray][@stdlib/ndarray/ctor]. By default, the index from which to begin searching is `0`.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/README.md:
```diff
@@ -0,0 +1,262 @@
+
+The function accepts the following options:
+
+-   **dtype**: output ndarray [data type][@stdlib/ndarray/dtypes]. Must be an "integer_index_and_generic" [data type][@stdlib/ndarray/dtypes].
```
```suggestion
-   **dtype**: output ndarray [data type][@stdlib/ndarray/dtypes]. Must be an integer or generic [data type][@stdlib/ndarray/dtypes].
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/README.md:
```diff
@@ -0,0 +1,262 @@
+
+-   **x**: input [ndarray][@stdlib/ndarray/ctor].
+-   **searchElement**: element in an input [ndarray][@stdlib/ndarray/ctor] for which to find an index. May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having a [data type][@stdlib/ndarray/dtypes] same as the data type of the input [ndarray][@stdlib/ndarray/ctor]. If provided a scalar value, the value is cast to the data type of the input [ndarray][@stdlib/ndarray/ctor]. If prov ...
+-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an `integer` or `generic` [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the complement of the shape defined by `o ...
```
```suggestion
-   **fromIndex**: index from which to begin searching (_optional_). May be either a scalar value or an [ndarray][@stdlib/ndarray/ctor] having an `integer` or `generic` [data type][@stdlib/ndarray/dtypes]. If provided an [ndarray][@stdlib/ndarray/ctor], the value must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the complement of the shape defined by `options.dims`. For example, given the input shape `[2, 3, 4]` and `options.dims=[0]`, an [ndarray][@stdlib/ndarray/ctor] containing the index from which to begin searching must have a shape which is [broadcast-compatible][@stdlib/ndarray/base/broadcast-shapes] with the shape `[3, 4]`. Similarly, when performing the operation over all elements in a provided input [ndarray][@stdlib/ndarray/ctor], an [ndarray][@stdlib/ndarray/ctor] containing the index from which to begin searching must be a zero-dimensional [ndarray][@stdlib/ndarray/ctor]. By default, the index from which to begin searching is `0`.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/README.md:
```diff
@@ -0,0 +1,262 @@
+
+## Notes
+
+-   Both functions iterate over [ndarray][@stdlib/ndarray/ctor] elements according to the memory layout of the input [ndarray][@stdlib/ndarray/ctor]. Accordingly, performance degradation is possible when operating over multiple dimensions of a large non-contiguous multi-dimensional input [ndarray][@stdlib/ndarray/ctor]. In such scenarios, one may want to copy an input [ndarray][@stdlib/ndarray/ct ...
```
Copy-paste error. We are not computing the cumulative sum.

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/assign.js:
```diff
@@ -0,0 +1,178 @@
+
+	nargs = arguments.length;
+	if ( nargs < 3 ) {
+		throw new TypeError( format( 'invalid argument. The function must be provided at least two arguments. Value: `%s`.', nargs ) );
```
```suggestion
		throw new TypeError( format( 'invalid argument. The function must be provided at least three arguments. Value: `%s`.', nargs ) );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/index-of/lib/assign.js:
```diff
@@ -0,0 +1,178 @@
+	} else {
+		throw new TypeError( format( 'invalid argument. Third argument must be either an ndarray or an integer. Value: `%s`.', fromIndex ) );
+	}
+
```
Missing `out` argument validation.

headlessNode replied:
Here also, notice the case `assign( x, searchElement, out, options )`  at L132, and also the case above `assign( x, searchElement, out )`. Since `fromIndex` is optional and can be at 3rd place, the issue is that `fromIndex` can also be `ndarrayLike` so consider the case `assign( x, searchElement, fromIndex_ndarray )`. How would we know if that ndarray is a `fromIndex` or `out`?

Based on that, I think, the check for `out` makes sense just before the call to `base`.

Also, as suggested by you, I tried to comeup with `ndarrayLike` checks at the top but that also is tricky because we need to first check if `fromIndex` is actually that or is it `out` & vice versa. So at the end, we will end up with same amount of `ndarraylike` checks as we have now.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/docs/repl.txt:
```diff
@@ -0,0 +1,126 @@
+        which is broadcast-compatible with the shape `[3, 4]`. Similarly
+        when performing the operation over all elements in a provided input
+        ndarray, an ndarray containing the index from which to begin searching
+        must be a zero-dimensional ndarray. By default, the value of the index
+        from which to begin searching is `0`.
```
```suggestion
        must be a zero-dimensional ndarray. By default, the index
        from which to begin searching is `0`.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/docs/repl.txt:
```diff
@@ -0,0 +1,126 @@
+        Function options.
+
+    options.dtype: string (optional)
+        Output array data type. Must be a "integer_index_and_generic" data type.
```
```suggestion
        Output array data type. Must be an integer or generic data type.
```

headlessNode committed:
docs: apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/index-of/README.md:
```diff
@@ -0,0 +1,262 @@
+
+## Notes
+
+-   Both functions iterate over [ndarray][@stdlib/ndarray/ctor] elements according to the memory layout of the input [ndarray][@stdlib/ndarray/ctor]. Accordingly, performance degradation is possible when operating over multiple dimensions of a large non-contiguous multi-dimensional input [ndarray][@stdlib/ndarray/ctor]. In such scenarios, one may want to copy an input [ndarray][@stdlib/ndarray/ct ...
```
```suggestion
-   Both functions iterate over [ndarray][@stdlib/ndarray/ctor] elements according to the memory layout of the input [ndarray][@stdlib/ndarray/ctor]. Accordingly, performance degradation is possible when operating over multiple dimensions of a large non-contiguous multi-dimensional input [ndarray][@stdlib/ndarray/ctor]. In such scenarios, one may want to copy an input [ndarray][@stdlib/ndarray/ctor] to contiguous memory before searching for an index.
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode committed:
fix: apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte committed:
docs: fix descriptions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
fix: refactor implementation to only support a single reduction dimension

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix example

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
fix: refactor implementation to only support a single reduction dimension

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix descriptions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix declarations and add notes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update README and fix option handling

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (changes requested):
I've updated the implementation. The only thing remaining are updating the tests. I suggest you take this on, as this will be a good exercise in both verifying my changes and understanding the changes made.

kgryte committed:
docs: update comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: main tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: assign tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: fix grammar in comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte committed:
bench: use search value which is out-of-range in order to guarantee worst case scenario

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: remove duplicate word

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update description

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
test: update tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
test: update tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
test: add output argument tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after some additional clean-up.

@headlessNode Re: tests. Notice how I updated the `fromIndex` test cases. Using a `fromIndex` of `0` is not informative, as that is the default value, so you were not actually testing whether a from index was supported.

Further notice how I added comments illustrating array shapes which makes interpreting expected values easier to infer.
